### PR TITLE
Issue #2855836 by mglaman, vasike: Coupon codes need unique value validation constraint

### PR DESCRIPTION
**Pull request by vasike**

### CODE REVIEW

File: modules/promotion/tests/src/Kernel/CouponValidationTest.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    // Add other coupon with the same code.
+    $coupon = Coupon::create([
+      'code' => $coupon_code,
+      'status' => TRUE,
+    ]);
+
+    $violations = $coupon->validate();
+    $this->assertEquals(count($violations), 1);
+    $this->assertEquals($violations[0]->getPropertyPath(), 'code');
+    $this->assertEquals($violations[0]->getMessage(), 'A coupon with code %value already exists. Enter a unique code.');
```
Comment (mglaman) [inline modules/promotion/tests/src/Kernel/CouponValidationTest.php:57]: Can we update the test to change the code value and run validations again? I'd like the test to prove we can still save the coupon with the validation now added. We have tests that should prove this, but I'd rather be explicit in our test coverage.

**Review by mglaman — CHANGES_REQUESTED** [review]
Code looks good to me. It's pretty much `\Drupal\Core\Validation\Plugin\Validation\Constraint\UniqueFieldValueValidator` but with specific message.

File: modules/promotion/tests/src/Kernel/CouponValidationTest.php (hunk 1 above)
Comment (vasike) [reply modules/promotion/tests/src/Kernel/CouponValidationTest.php:57]: done

Comment (mglaman) [reply modules/promotion/tests/src/Kernel/CouponValidationTest.php:57]: 💃 yay

**Review by mglaman — APPROVED** [review]

**Comment by mglaman** [comment]
@bojanz this looks good to me, now. My main Q is just should we use custom validator or reuse core one. But seems better to be verbose.

**Comment by bojanz** [comment]
Committed in 0a7d06cb05855d41f9af595d380841d25a21c695b.

**Comment by bojanz** [comment]
Committed in 0a7d06cb05855d41f9af595d380841d25a21c695b.

**Pull request closed without merging**